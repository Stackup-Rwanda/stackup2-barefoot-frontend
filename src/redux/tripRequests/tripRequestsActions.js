import axios from 'axios';
import {
  FETCH_TRIP_REQUESTS_REQUEST,
  FETCH_TRIP_REQUESTS_SUCCESS,
  FETCH_TRIP_REQUESTS_FAILURE,
} from './tripRequestsTypes';

export const fetchTripRequestsRequest = () => ({
  type: FETCH_TRIP_REQUESTS_REQUEST,
});

export const fetchTripRequestsSuccess = (tripRequests) => ({
  type: FETCH_TRIP_REQUESTS_SUCCESS,
  payload: tripRequests,
});

export const fetchTripRequestsFailure = (error) => ({
  type: FETCH_TRIP_REQUESTS_FAILURE,
  payload: error,
});

export const api = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
  headers: {
    Authorization: `Bearer ${process.env.REACT_APP_SOME_TOKEN}`,
    'Content-Type': 'application/json',
  },
  method: 'GET',
});

export const fetchTripRequests = () => async (dispatch) => {
  dispatch(fetchTripRequestsRequest);
  await api.get('/trips')
    .then((res) => {
      dispatch(fetchTripRequestsSuccess(res.data.data.foundRequests));
    }).catch((err) => {
      if (err.response.status === 400) {
        const errorMessage = err.response.data.error;
        dispatch(fetchTripRequestsFailure(errorMessage));
      } else {
        const tripRequests = [];
        dispatch(fetchTripRequestsSuccess(tripRequests));
      }
    });
};
