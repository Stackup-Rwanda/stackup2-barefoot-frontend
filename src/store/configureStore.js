import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import reducer from '../redux/login/reducer';

const configureStore = () => {
  const middlewares = [thunk];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const enhancers = [middlewareEnhancer, composeWithDevTools()];
  const composedEnhancers = compose(...enhancers);

  const store = createStore(reducer, undefined, composedEnhancers);

  return store;
};

export default configureStore;
