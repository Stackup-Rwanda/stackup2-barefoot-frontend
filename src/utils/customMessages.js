export default {
  welcomeToBarefootNomad: `Hello there, welcome to the new world.
  This is the frontend of the BAREFOOT NOMAD APP developed using with
  react the-champions team!`,
};
