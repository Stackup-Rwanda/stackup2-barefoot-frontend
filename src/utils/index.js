/* eslint-disable import/prefer-default-export */
import { createStore, applyMiddleware } from 'redux';
import rootReducer from '../redux/rootReducer';
import { middlewares } from '../redux/store';

export const mockStore = (initialState) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, initialState);
};
