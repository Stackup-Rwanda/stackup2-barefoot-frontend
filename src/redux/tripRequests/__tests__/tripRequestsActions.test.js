/* eslint-disable no-undef */
import moxios from 'moxios';
import configureStore from 'redux-mock-store';
import * as actions from '../tripRequestsActions';
import { mockStore } from '../../../utils';

const testStore = configureStore();
const store = testStore();

describe('TripRequests actions', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches the FETCH_TRIP_REQUESTS_REQUEST action', () => {
    store.dispatch(actions.fetchTripRequestsRequest());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches the FETCH_TRIP_REQUESTS_SUCCESS action', () => {
    store.dispatch(actions.fetchTripRequestsSuccess());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches the FETCH_TRIP_REQUESTS_FAILURE action', () => {
    store.dispatch(actions.fetchTripRequestsFailure());
    expect(store.getActions()).toMatchSnapshot();
  });

  // it('Mocks the fetchTripRequests special action', async () => {
  //   actions.fetchTripRequests = jest.fn(() =>
  // store.dispatch(actions.fetchTripRequestsRequest()));
  //   actions.fetchTripRequests();
  //   store.dispatch(actions.fetchTripRequests());
  //   expect(actions.fetchTripRequests).toHaveBeenCalled();
  // });
});

describe('FetchTripRequests action', () => {
  beforeEach(() => {
    moxios.install(actions.api);
  });

  afterEach(() => {
    moxios.uninstall(actions.api);
  });

  it('Updates store successfully', () => {
    const expectedState = [
      {
        id: 1,
        trips: [
          {
            travelDate: '2020-09-15',
            travelFrom: 'Kigali',
            travelTo: 'Durban',
          },
        ],
        travelReason: 'Africa Tech Summit',
        accommodation: true,
        status: 'pending',
      },
      {
        id: 2,
        trips: [
          {
            travelDate: '2020-09-10',
            travelFrom: 'Kigali',
            travelTo: 'Dubai',
          },
        ],
        travelReason: 'Africa Tech Summit',
        accommodation: true,
        status: 'pending',
      },
    ];
    const store = mockStore();

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: expectedState,
      });
      request.reject({
        status: 400,
        response: {
          status: 400,
          data: {
            error: 'Invalid token',
          },
        },
      });
    });

    return store.dispatch(actions.fetchTripRequests()).then(() => {
      const newState = store.getState();
      expect(newState.requests).toBe(expectedState);
    }).catch(() => {
      const newState = store.getState();
      expect(newState.tripRequests.tripRequests).toEqual([]);
    });
  });
});
