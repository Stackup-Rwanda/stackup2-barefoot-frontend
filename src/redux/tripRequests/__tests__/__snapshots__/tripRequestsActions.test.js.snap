// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripRequests actions Dispatches the FETCH_TRIP_REQUESTS_FAILURE action 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "FETCH_TRIP_REQUESTS_FAILURE",
  },
]
`;

exports[`TripRequests actions Dispatches the FETCH_TRIP_REQUESTS_REQUEST action 1`] = `
Array [
  Object {
    "type": "FETCH_TRIP_REQUESTS_REQUEST",
  },
]
`;

exports[`TripRequests actions Dispatches the FETCH_TRIP_REQUESTS_SUCCESS action 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "FETCH_TRIP_REQUESTS_SUCCESS",
  },
]
`;
